HatchGenerator/Freecad_Hatch_Generator.py

```diff
@@ -0,0 +1,4355 @@
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        # Non‑modal dialog flag
+        self.setWindowFlags(self.windowFlags() | QtCore.Qt.Window)
+        if getattr(QtWidgets, 'QApplication', None):
+            try:
+                self.setWindowIcon(FreeCADGui.getMainWindow().windowIcon())
+            except Exception:
+                pass
+
+        self.setWindowTitle("Custom Hatch with Variation Modes")
+        self.resize(360, 600)  # Made a bit taller
+        self.doc = FreeCAD.ActiveDocument
+        if not self.doc:
+            raise Exception("No active document found.")
+
+        self.current_temp_preview_name = None
+        self.hatch_created = False
+        self.masterObjectList = self.getAllObjectsClassified()
+        self.last_manual_scale = None
+
+        self.initWidgets()
+        self.setup_ui()
+        self.initialPatternSetup()
+
+    def initWidgets(self):
+        """
+        Create all widget objects.
+        """
+
+        # ============= Main Tab Elements =============
+        self.baseLabel = QtWidgets.QLabel("Base Shape:")
+        self.baseCombo = QtWidgets.QComboBox()
+        self.baseSearchLabel = QtWidgets.QLabel("Search:")
+        self.baseSearchField = QtWidgets.QLineEdit()
+        self.baseTypeFilterLabel = QtWidgets.QLabel("Type Filter:")
+        self.baseTypeFilterCombo = QtWidgets.QComboBox()
+        for cat in ["All", "Sketch", "PartFeature", "Compound", "Other"]:
+            self.baseTypeFilterCombo.addItem(cat)
+        # Pick base shape button
+        self.pickBaseBtn = QtWidgets.QPushButton("Pick Base Shape from Selection")
+        # Pick Custom Pattern Shape button
+        self.pickCustomBtn = QtWidgets.QPushButton("Pick Custom Pattern from Selection")
+        self.pickCustomBtn.clicked.connect(self.pickCustomPattern)
+
+        # Pick Subtraction Objects button
+        self.pickSubBtn = QtWidgets.QPushButton("Pick Subtractions from Selection")
+        self.pickSubBtn.clicked.connect(self.pickSubtractions)
+
+        # Pick Base Tile button
+        self.pickTileBtn = QtWidgets.QPushButton("Pick Base Tile from Selection")
+        self.pickTileBtn.clicked.connect(self.pickBaseTile)
+        self.pickBaseBtn.clicked.connect(self.pickBaseShape)
+
+
+        # === [TWEAKED / NEW UI ELEMENTS ADDED] ===
+        # BaseObjects multi-list
+        self.baseObjectsLabel = QtWidgets.QLabel("Additional Base Objects:")
+        self.baseObjectsList = QtWidgets.QListWidget()
+        self.baseObjectsList.setSelectionMode(QtWidgets.QAbstractItemView.MultiSelection)
+
+        # Pattern Source
+        self.patternSourceLabel = QtWidgets.QLabel("Pattern Source:")
+        self.patternSourceCombo = QtWidgets.QComboBox()
+        self.patternSourceCombo.addItems(["Built-in", "Custom"])
+
+        self.builtinPatternLabel = QtWidgets.QLabel("Built-in Pattern:")
+        self.builtinPatternCombo = QtWidgets.QComboBox()
+        self.builtinPatternCombo.addItems([
```

Comment:
It'll be more future-proof to save the list of supported hatches in a global variable (or have patterns as separate files, as suggested above).